File: DotNet/RestSample/QiClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

using Microsoft.IdentityModel.Clients.ActiveDirectory;
using Newtonsoft.Json;

namespace RestSample
{
    public class QiClient
    {
        private HttpClient _httpClient;
        private string _baseUrl;

        // VERY IMPORTANT: edit the following values to reflect the authorization items you were given

        static string _resource = "PLACEHOLDER_REPLACE_WITH_RESOURCE";
        static string _authority = "PLACEHOLDER_REPLACE_WITH_AUTHORITY";
        static string _appId = "PLACEHOLDER_REPLACE_WITH_USER_ID";
        static string _appKey = "PLACEHOLDER_REPLACE_WITH_USER_SECRET";

        // Azure AD authentication related
        private static AuthenticationContext _authContext = null;

        // REST API url strings
        private string _tenantsBase = @"Qi/Tenants";
        private string _typesBase = @"Qi/Types";
        private string _streamsBase = @"Qi/Streams";
        private string _insertSingle = @"/Data/InsertValue";
        private string _insertMultiple = @"/Data/InsertValues";
        private string _getTemplate = @"/Data/GetWindowValues?startIndex={0}&endIndex={1}";
        private string _updateSingle = @"/Data/UpdateValue";
        private string _updateMultiple = @"/Data/UpdateValues";
        private string _removeSingleTemplate = @"/{0}/Data/RemoveValue?index={1}";
        private string _removeMultipleTemplate = @"/{0}/Data/RemoveWindowValues?startIndex={1}&endIndex={2}";

        // void error formats
        private string _createError = "Failed to create {0} with Id = {1}";
        private string _updateError = "Failed to update {0} with Id = {1}";
        private string _deleteError = "Failed to delete event from stream {0} with index = {1}";
        private string _deleteMultipleError = "Failed to delete events from stream {0} with indices {1}";

        public QiClient(string baseUrl)
        {
            _httpClient = new HttpClient();
            _baseUrl = baseUrl;
            if (_baseUrl.Substring(_baseUrl.Length - 1).CompareTo(@"/") != 0)
                _baseUrl = _baseUrl + "/";

            _httpClient.BaseAddress = new Uri(_baseUrl); 
            _httpClient.Timeout = new TimeSpan(0, 0, 30);
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }
        #region Metadata methods
        // TODO retract when provisioning is complete
        public async Task CreateTenant(string tenantId)
        {
            QiTenant tenant = new QiTenant(tenantId);
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _tenantsBase),
                Method = HttpMethod.Post,
            };

            //msg.Headers.Authorization = "Bearer: x";
            
            string content = JsonConvert.SerializeObject(tenant);
            msg.Content = new StringContent(content, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (!response.IsSuccessStatusCode)
            {
                throw new QiError(response.StatusCode, "Failed to create tenant with Id = " + tenantId);
            }
        }

        public async Task DeleteTenant(string tenantId)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _tenantsBase + @"/" + tenantId),
                Method = HttpMethod.Delete
            };

            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (!response.IsSuccessStatusCode)
            {
                throw new QiError(response.StatusCode, "Failed to create tenant with Id = " + tenantId);
            }
        }

        /// <summary>
        /// Create a stream on the target Qi Service
        /// </summary>
        /// <param name="streamDef">QiStream object with name, Id, type</param>
        /// <returns>void</returns>
        public async Task<string> CreateStream(QiStream streamDef)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase),
                Method = HttpMethod.Post,
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            string content = JsonConvert.SerializeObject(streamDef);
            msg.Content = new StringContent(content, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (!response.IsSuccessStatusCode)
            {
                throw new QiError(response.StatusCode, "Error creating Stream with id " + streamDef.Id);
            }
            else
            {
                return await response.Content.ReadAsStringAsync();
            }
        }

        /// <summary>
        /// Deletes the stream whose Id is passed
        /// </summary>
        /// <param name="streamId">Id of the QiStream to be deleted</param>
        /// <returns>void</returns>
        public async Task DeleteStream(string streamId)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId),
                Method = HttpMethod.Delete,
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            await SendAndRespondVoid(msg, _createError, "stream", streamId);
        }

        /// <summary>
        /// Creates a QiType in the remote Qi Service.  QiTypes are required by QiStreams and cannot be deleted while so referenced.
        /// </summary>
        /// <param name="typeDef">QiType object defining a user-defined type with properties</param>
        /// <returns>void</returns>
        public async Task<string> CreateType(QiType typeDef)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _typesBase),
                Method = HttpMethod.Post,
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            string content = JsonConvert.SerializeObject(typeDef);
            msg.Content = new StringContent(content, Encoding.UTF8, "application/json");

            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (!response.IsSuccessStatusCode)
            {
                throw new QiError(response.StatusCode, "Error creating Type with id " + typeDef.Id);
            }
            else
            {
                return await response.Content.ReadAsStringAsync();
            }
        }

        /// <summary>
        /// Deletes a QiType in the remote Qi Service
        /// </summary>
        /// <param name="typeId">Id of the QiType to delete</param>
        /// <returns>void</returns>
        public async Task DeleteType(string typeId)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _typesBase +@"/" + typeId),
                Method = HttpMethod.Delete,
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

           await  SendAndRespondVoid(msg, _deleteError, "type", typeId);
        }
        #endregion

        #region Create Methods for Data (Insert)

        /// <summary>
        /// Creates a single instance of a measured event in the named Qi Stream provided the object conforms to the type of the stream
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="evt">JSON serialization of the object to store</param>
        /// <returns></returns>
        public async Task CreateEvent(string streamId, string evt)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId + _insertSingle),
                Method = HttpMethod.Post
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            msg.Content = new StringContent(evt, Encoding.UTF8, "application/json");
            await SendAndRespondVoid(msg, _createError, "data event", streamId);
        }

        /// <summary>
        /// Creates multiple events in the identified Qi Stream
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="evt">JSON serialization of an array of events</param>
        /// <returns>void</returns>
        public async Task CreateEvents(string streamId, string evt)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId + _insertMultiple),
                Method = HttpMethod.Post
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            msg.Content = new StringContent(evt, Encoding.UTF8, "application/json");
            await SendAndRespondVoid(msg, _createError, "data events", streamId);
        }

        #endregion

        #region Retrieve Methods for Data
        /// <summary>
        /// Gets all values in the range identified by the start and end indices
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="startIndex">string representation of a value related to the streams key; for a DateTime key, an ISO 8601 time reference</param>
        /// <param name="endIndex">string representaion of a value related to the stream's key denoting the end of the desired range</param>
        /// <returns>JSON serialized array of events</returns>
        public async Task<string> GetWindowValues(string streamId, string startIndex, string endIndex)
        {   
            string getClause = string.Format(_getTemplate, startIndex, endIndex);
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId + getClause),
                Method = HttpMethod.Get
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (response.IsSuccessStatusCode)
            {
                string jsonResults = await response.Content.ReadAsStringAsync();
                return jsonResults;
            }
            else
            {
                throw new QiError(response.StatusCode, "Error getting windows values: " + response.ReasonPhrase);
            }
        }

        #endregion

        #region Update Methods for Data
        /// <summary>
        /// Updates a single event if found
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="evt">JSON serialization of the updated event</param>
        /// <returns></returns>
        public async Task UpdateValue(string streamId, string evt)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId + _updateSingle),
                Method = HttpMethod.Put
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            msg.Content = new StringContent(evt, Encoding.UTF8, "application/json");
            await SendAndRespondVoid(msg, _updateError, "data event", evt);
        }

        /// <summary>
        /// Updates an array of events in a stream (replaces an event with the same key value)
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="evts">JSON serialization of an array of updated events</param>
        /// <returns>void</returns>
        public async Task UpdateValues(string streamId, string evts)
        {
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(_baseUrl + _streamsBase + @"/" + streamId + _updateMultiple),
                Method = HttpMethod.Put
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            msg.Content = new StringContent(evts, Encoding.UTF8, "application/json");
            await SendAndRespondVoid(msg, _updateError, "data events", streamId);
        }

        #endregion

        /// <summary>
        /// removes a single value from the stream if the index is found
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="index">value of the key of the event to be deleted (for DateTime, ISO 8601 string)</param>
        /// <returns>void</returns>
        public async Task RemoveValue(string streamId, string index)
        {
            string uri = _baseUrl + _streamsBase + string.Format(_removeSingleTemplate, streamId, index);
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(uri),
                Method = HttpMethod.Delete
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            await SendAndRespondVoid(msg, _deleteError, streamId, index);
        }

        /// <summary>
        /// Removes all events within a range in a stream
        /// </summary>
        /// <param name="streamId">stream identifier</param>
        /// <param name="startIndex">start key value for the range</param>
        /// <param name="endIndex">end key value for the range</param>
        /// <returns>void</returns>
        public async Task RemoveWindowValues(string streamId, string startIndex, string endIndex)
        {
            string uri = _baseUrl + _streamsBase + string.Format(_removeMultipleTemplate, streamId, startIndex, endIndex);
            HttpRequestMessage msg = new HttpRequestMessage
            {
                RequestUri = new Uri(uri),
                Method = HttpMethod.Delete
            };

            string token = AcquireAuthToken();
            msg.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            await SendAndRespondVoid(msg, _deleteMultipleError, streamId, startIndex + ", " + endIndex);
        }

        private async Task SendAndRespondVoid(HttpRequestMessage msg, string template, string entity, string id)
        {
            HttpResponseMessage response = await _httpClient.SendAsync(msg);
            if (!response.IsSuccessStatusCode)
            {
                throw new QiError(response.StatusCode, string.Format(template, entity, id));
            }

        }

        static protected string AcquireAuthToken()
        {
            if (_authContext == null)
            {
                _authContext = new AuthenticationContext(_authority);
            }

            // tokens expire after a certain period of time
            // You can check this with the ExpiresOn property of AuthenticationResult, but it is not necessary.
            // ADAL maintains an in-memory cache of tokens and transparently refreshes them as needed
            try
            {
                ClientCredential userCred = new ClientCredential(_appId, _appKey);
                AuthenticationResult authResult = _authContext.AcquireToken(_resource, userCred);
                return authResult.AccessToken;
            }
            catch (AdalException)
            {
                return string.Empty;
            }


        }

    }
}

File: Java/samples/src/main/java/samples/QiClient.java
package samples;

import java.net.*;
import java.util.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.io.*;


import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.microsoft.aad.adal4j.AuthenticationContext;
import com.microsoft.aad.adal4j.AuthenticationResult;
import com.microsoft.aad.adal4j.ClientCredential;


public class QiClient {

	Gson mGson =  null; 
	private String baseUrl = null; 
	private String tenantName = null;
	static String _resource = "https://pihomemain.onmicrosoft.com/historian";
	static String _authority = "https://login.windows.net/qimaininternal.onmicrosoft.com";
	
	static String _appId = "Please enter your appId here";
	static String _appKey = "Please enter your appKey here";



	// REST API url strings
	private String tenantsBase = "/Qi/Tenants";
	private String typesBase = "/Qi/Types";
	private String streamsBase = "/Qi/Streams";
	private String insertSingle = "/Data/InsertValue";
	private String insertMultiple = "/Data/InsertValues";
	private String getTemplate = "/Data/GetWindowValues?";
	private String updateSingle = "/Data/UpdateValue";
	private String updateMultiple = "/Data/UpdateValues";
	private String removeSingleTemplate = "/{0}/Data/RemoveValue?index={1}";
	private String removeMultipleTemplate = "/{0}/Data/RemoveWindowValues?startIndex={1}&endIndex={2}";
	private static AuthenticationContext _authContext = null;
	private static ExecutorService service = null;
	private static AuthenticationResult result = null;

	// void error formats fixme
	//private String _createError = "Failed to create {0} with Id = {1}";


	public static  java.net.HttpURLConnection getConnection(URL url, String method){
		java.net.HttpURLConnection urlConnection = null;

		AuthenticationResult token = AcquireAuthToken();
		try{
			urlConnection = (java.net.HttpURLConnection) url.openConnection();
			urlConnection.setRequestMethod(method);
			urlConnection.setUseCaches(false);
			urlConnection.setConnectTimeout(10000);
			urlConnection.setReadTimeout(10000);
			urlConnection.setRequestProperty("Content-Type", "application/json");
			urlConnection.setRequestProperty( "Authorization", token.getAccessTokenType()+ " "+ token.getAccessToken());
			if (method == "POST" || method == "PUT" || method == "DELETE"){  	
				urlConnection.setChunkedStreamingMode(0);
				urlConnection.setDoOutput(true);     
			}else if(method == "GET"){
			}


		}catch(SocketTimeoutException e){
			e.getMessage();

		}

		catch (ProtocolException e){

			e.getMessage();

		} catch (IllegalStateException e) {
			e.getMessage();

		}catch(Exception e)  {
			e.printStackTrace();

		}

		return urlConnection;
	}



	public QiClient(String baseUrl)
	{
		mGson = new GsonBuilder().registerTypeAdapter(GregorianCalendar.class, new UTCDateTypeAdapter()).setDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").create();
		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		this.baseUrl =  baseUrl;
		
		try
		{
			url = new URL(this.baseUrl);		
			urlConnection = getConnection(url, "POST");
			urlConnection.setDoOutput(true);
			urlConnection.setRequestMethod("POST");

		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}catch (ProtocolException e){

			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}





	}
	public String CreateType(QiType typeDef)
	{

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		String inputLine;
		StringBuffer response = new StringBuffer();

		try
		{
			url = new URL(baseUrl + typesBase );

			urlConnection = getConnection(url,"POST");
		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}        
		catch (Exception e) {
			e.printStackTrace();
		}



		try{
			String body = mGson.toJson(typeDef);           
			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(body);
			writer.close();

			int HttpResult = urlConnection.getResponseCode();




			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("type creation request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{
				throw new QiError(urlConnection, "Type creation failed");
			}

			BufferedReader in = new BufferedReader(
					new InputStreamReader(urlConnection.getInputStream()));

			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}



		}catch (Exception e){

			e.printStackTrace();

		}

		return response.toString();


	}


	public String CreateStream(QiStream streamDef)
	{

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		String inputLine;
		StringBuffer response = new StringBuffer();


		try
		{
			url = new URL(baseUrl + streamsBase );

			urlConnection = getConnection(url,"POST");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();

		}catch (Exception e) {
			e.printStackTrace();
		}

		try
		{    
			String body = mGson.toJson(streamDef);           
			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(body);
			writer.close();

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("stream creation request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				throw new QiError(urlConnection, "Stream creation failed");
			}



			BufferedReader in = new BufferedReader(
					new InputStreamReader(urlConnection.getInputStream()));


			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}
			in.close();


		}catch (Exception e){

			e.printStackTrace();

		}

		return response.toString();

	}



	public void CreateEvent(String streamId, String evt)
	{
		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;

		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + insertSingle);

			urlConnection = getConnection(url,"POST");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}      
		catch (Exception e) {
			e.printStackTrace();
		}



		try
		{


			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(evt);
			writer.close();

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("Event creation request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{
				throw new QiError(urlConnection, "Event creation failed");

			}



		}catch (Exception e){



		}
	}



	public void CreateEvents(String streamId, String json) throws QiError {

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		int HttpResult = 0; 
		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + insertMultiple);

			urlConnection = getConnection(url,"POST");
		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}      
		catch (Exception e) {
			e.printStackTrace();
		}


		try
		{

			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(json);
			writer.close();

			HttpResult = urlConnection.getResponseCode();

		}catch (Exception e){

			e.printStackTrace();

		}
		if (HttpResult == HttpURLConnection.HTTP_OK)
		{
			System.out.println("Events creation request succeded");
		}


		if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
		{
			throw new QiError(urlConnection, "Events creation failed");

		}

	}



	public String GetWindowValues (String streamId, String startIndex, String endIndex)throws QiError
	{   
		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;

		String inputLine;
		StringBuffer jsonResults = new StringBuffer();


		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + getTemplate +  "startIndex=" + startIndex + "&" + "endIndex=" + endIndex);

			urlConnection = getConnection(url,"GET");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}
            

		try{
			int HttpResult = urlConnection.getResponseCode();

			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("GetWindowValues request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{
				throw new QiError(urlConnection, "GetWindowValues request failed");
			}


			BufferedReader in = new BufferedReader(
					new InputStreamReader(urlConnection.getInputStream()));


			while ((inputLine = in.readLine()) != null) {
				jsonResults.append(inputLine);
			}
			in.close();


		}catch (Exception e){

			e.printStackTrace();

		}

		return jsonResults.toString();




	}



	public void updateValue(String streamId, String json)throws QiError {
		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;


		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + updateSingle);

			urlConnection = getConnection(url,"PUT");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}

		try
		{

			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(json);
			writer.close();

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("UpdateValue request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				throw new QiError(urlConnection, "GetWindowValues request failed");
			}

		}catch (Exception e){

			e.printStackTrace();

		}




	}


	public void updateValues(String streamId, String json) {
		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;


		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + updateMultiple);

			urlConnection = getConnection(url,"PUT");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}




		try
		{

			OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
			OutputStreamWriter writer = new OutputStreamWriter(out);
			writer.write(json);
			writer.close();

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("Update Values request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				throw new QiError(urlConnection, "Update Values request failed");
			}

		}catch (Exception e){

			e.printStackTrace();

		}


	}



	public void removeValue(String streamId, String index) throws QiError{

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;



		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + "/" + "/Data/RemoveValue?index=" + index);

			urlConnection = getConnection(url,"DELETE");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}

		try
		{

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("remove Value request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{
				throw new QiError(urlConnection, "Remove value request failed");
			}
		}catch (Exception e){

			e.printStackTrace();

		}


	}



	public void removeWindowValues(String streamId, String startIndex, String endIndex) {

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;

		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId + "/Data/RemoveWindowValues?startIndex=" + startIndex + "&" + "endIndex=" + endIndex );

			urlConnection = getConnection(url,"DELETE");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}




		try
		{


			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("removeWindowValues request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				throw new QiError(urlConnection, "Remove windows value request failed");
			}

		}catch (Exception e){

			e.printStackTrace();

		}

	}



	public void deleteStream(String streamId) {

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;

		try
		{
			url = new URL(baseUrl + streamsBase + "/" + streamId );
			urlConnection = getConnection(url,"DELETE");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}



		try
		{

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("deleteStream request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				throw new QiError(urlConnection, "Delete  Stream failed");
			}

		}catch (Exception e){

			e.printStackTrace();

		}
	}



	public void deleteType(String typeId) {

		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		try
		{
			url = new URL(baseUrl + typesBase + "/" + typeId );
			urlConnection = getConnection(url,"DELETE");


		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException");
		}catch (IllegalStateException e) {
			e.getMessage();
		}         
		catch (Exception e) {
			e.printStackTrace();
		}

		try
		{

			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("deleteType request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{
				throw new QiError(urlConnection, "Delete type failed");

			}

		}catch (Exception e){

			e.printStackTrace();

		}

	}



	public void deleteTenant(String tenantId) {


		java.net.URL url = null;
		java.net.HttpURLConnection urlConnection = null;
		String inputLine;
		try
		{
			url = new URL(baseUrl + tenantsBase + "/" + tenantId );
		}
		catch (MalformedURLException mal)
		{
			System.out.println("MalformedURLException for delete tenant");
		}


		try
		{
			urlConnection = (java.net.HttpURLConnection) url.openConnection();
			urlConnection.setDoOutput(true);
			urlConnection.setChunkedStreamingMode(0);
			urlConnection.setRequestMethod("DELETE");
			urlConnection.setUseCaches(false);
			urlConnection.setConnectTimeout(10000);
			urlConnection.setReadTimeout(10000);
			urlConnection.setRequestProperty("Content-Type", "application/json");
			urlConnection.setRequestProperty("QiTenant",tenantName );  



			int HttpResult = urlConnection.getResponseCode();
			if (HttpResult == HttpURLConnection.HTTP_OK)
			{
				System.out.println("deleteTenant request succeded");
			}


			if (HttpResult != HttpURLConnection.HTTP_OK && HttpResult != HttpURLConnection.HTTP_CREATED)
			{

				System.out.println("deleteTenant request failed");
				BufferedReader in = new BufferedReader(
						new InputStreamReader(urlConnection.getErrorStream()));


				while ((inputLine = in.readLine()) != null) {
					System.out.println(inputLine);
				}

				in.close();
			}

		}catch (Exception e){



		}


	}



	static protected AuthenticationResult AcquireAuthToken()
	{

		service = Executors.newFixedThreadPool(1);
		try
		{
			if (_authContext == null)
			{
				_authContext = new AuthenticationContext(_authority, true , service);
			}

			// tokens expire after a certain period of time
			// You can check this with the ExpiresOn property of AuthenticationResult, but it is not necessary.
			// ADAL maintains an in-memory cache of tokens and transparently refreshes them as needed

			ClientCredential userCred = new ClientCredential(_appId, _appKey);
			Future<AuthenticationResult> authResult = _authContext.acquireToken(_resource, userCred, null);

			//AcquireToken(_resource, userCred);
			result = authResult.get();


		}
		catch (Exception e)
		{


		}finally {
			service.shutdown();
		}
		return result;  

	}






}
